src/Percentiler.h

```diff
@@ -0,0 +1,97 @@
+#ifndef GCP_PERCENTILER_H
+#define GCP_PERCENTILER_H
+
+#include <iostream>
+#include <map>
+#include <vector>
+#include <algorithm>
+#include <stdexcept>
+
+#include "custom_types.h"
+
+
+class NotEnoughData : public std::runtime_error {
+  using std::runtime_error::runtime_error;
+};
+
+/**
+ * Class responsible for assigning confidence percentiles to raw genotype confidences.
+ * This class is decoupled from the others (to satisfy https://github.com/leoisl/GCP/issues/6)
+ */
+class Percentiler {
+public:
+  /**
+   * Builds a GenotypeConfidencePercentiler from a vector of simulated genotype confidences
+   */
+  explicit Percentiler(const std::vector<GenotypeConfidence> &input_entries) :
+          entries(input_entries),
+          min_entry(*std::min_element(input_entries.begin(), input_entries.end())),
+          max_entry(*std::max_element(input_entries.begin(), input_entries.end())) {
+
+      bool enough_data = input_entries.size() >= 2;
+      if (not enough_data) {
+          throw NotEnoughData("Please provide at least two simulated genotype confidences.");
+      }
+  }
+
+
+  /**
+   * Get the confidence percentile given a genotype confidence.
+   */
+  GenotypePercentile get_confidence_percentile(GenotypeConfidence query) {
+      auto lo = std::lower_bound(entries.begin(), entries.end(), query);
+      if (lo == entries.end()) return 100.0;
+      else if (*lo == query){
+          auto hi = std::upper_bound(entries.begin(), entries.end(), query);
+          if (lo == hi - 1) return iterator_to_percentile(lo);
+          // Case: multiple identical entries, take average
+          auto lo_percentile = iterator_to_percentile(lo);
+          auto hi_percentile = iterator_to_percentile(--hi);
+          return lo_percentile + (hi_percentile - lo_percentile) / 2;
+      }
+      else {
+          if (lo == entries.begin()) return 0.0;
```

Comment:
i prefer avoiding implicit else. the else would still be there implicitly right? its just that we return so it means it doesnt have to be. but the fact that its there means you think back to why, and it tells you that `*lo != query`

Reply:
Not sure I understand, I also did not state myself well. I meant to move this if just after or before `if (lo == entries.end()) return 100.0;`, so the conditions where we test for the limits are grouped, and we know that after these two conditions are tested, we are testing a query value inside the limits. Also would be nice to change all ifs so that they don't have a condition inside, but just test the value of a self documented boolean variable